client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -291,44 +286,82 @@ public final void addReconnectListener(final ReconnectListener<HonoConnection> l
-        } else if (connecting.get()) {
+        } else {
+            resultHandler.handle(Future.failedFuture(
+                    new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "not connected")));
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final Future<Void> isConnected(final long waitForCurrentConnectAttemptTimeout) {
+        return executeOnContext(result -> checkConnected(result, waitForCurrentConnectAttemptTimeout));
+    }
+
+    private void checkConnected(final Handler<AsyncResult<Void>> resultHandler, final long waitForCurrentConnectAttemptTimeout) {
+        if (isConnectedInternal()) {
+            resultHandler.handle(Future.succeededFuture());
+        } else if (waitForCurrentConnectAttemptTimeout > 0 && checkConnectedPromises.get() != null) {
             // connect attempt in progress - let its completion complete the resultHandler here
             log.debug("connection attempt to server [{}:{}] in progress, connection check will be completed with its result",
                     connectionFactory.getHost(), connectionFactory.getPort());
-            final Promise<Void> promiseToAdd = Promise.promise();
-            promiseToAdd.future().setHandler(resultHandler);
-            // atomically add the promise to checkConnectedPromises - but only if checkConnectedPromises hasn't been cleared already
-            final List<Promise<Void>> newCheckConnectedPromises = checkConnectedPromises
-                    .accumulateAndGet(Collections.singletonList(promiseToAdd), (existing, toAdd) -> {
-                        // no modification of the existing list done here, keeping the accumulatorFunction function side-effect free as required
-                        if (existing == null) {
-                            return null;
-                        }
-                        final List<Promise<Void>> promises = new ArrayList<>(existing.size() + 1);
-                        promises.addAll(existing);
-                        promises.add(toAdd.get(0));
-                        return promises;
-                    });
-            if (newCheckConnectedPromises == null) {
-                // checkConnectedPromises wasn't updated with our promiseToAdd because the checkConnectedPromises were cleared in between
-                resultHandler.handle(Future.failedFuture(
-                        new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "not connected")));
+            final CheckConnectedPromise promiseToAdd = new CheckConnectedPromise(resultHandler);
+            if (!addToCheckConnectedPromises(promiseToAdd)) {
+                // checkConnectedPromises has been cleared in between
+                checkConnected(resultHandler);
+            } else {
+                // promise added to list, now use a timer to ensure that we wait no more than the given timeout
+                promiseToAdd.setTimerId(vertx.setTimer(waitForCurrentConnectAttemptTimeout, id -> {
```

Comment:
shouldn't this better be encapsulated in `CheckConnectPromise` (which might be renamed to `ExpiringPromise` according to its purpose)?